ddtrace/debugging/_origin/span.py

```diff
@@ -0,0 +1,248 @@
+from dataclasses import dataclass
+from functools import partial as λ
+from itertools import count
+from pathlib import Path
+import sys
+from threading import current_thread
+from types import FrameType
+from types import FunctionType
+import typing as t
+import uuid
+
+import attr
+
+import ddtrace
+from ddtrace import config
+from ddtrace._trace.processor import SpanProcessor
+from ddtrace.debugging._debugger import Debugger
+from ddtrace.debugging._probe.model import DEFAULT_CAPTURE_LIMITS
+from ddtrace.debugging._probe.model import LiteralTemplateSegment
+from ddtrace.debugging._probe.model import LogFunctionProbe
+from ddtrace.debugging._probe.model import LogLineProbe
+from ddtrace.debugging._probe.model import ProbeEvaluateTimingForMethod
+from ddtrace.debugging._signal.collector import SignalContext
+from ddtrace.debugging._signal.snapshot import Snapshot
+from ddtrace.ext import EXIT_SPAN_TYPES
+from ddtrace.internal import compat
+from ddtrace.internal import core
+from ddtrace.internal.packages import is_user_code
+from ddtrace.internal.safety import _isinstance
+from ddtrace.internal.utils.inspection import linenos
+from ddtrace.internal.wrapping.context import WrappingContext
+from ddtrace.span import Span
+
+
+def frame_stack(frame: FrameType) -> t.Iterator[FrameType]:
+    _frame: t.Optional[FrameType] = frame
+    while _frame is not None:
+        yield _frame
+        _frame = _frame.f_back
+
+
+@attr.s
+class EntrySpanProbe(LogFunctionProbe):
+    __span_class__ = "entry"
+
+    @classmethod
+    def build(cls, name: str, module: str, function: str) -> "EntrySpanProbe":
+        message = f"{cls.__span_class__} span info for {name}, in {module}, in function {function}"
+
+        return cls(
+            probe_id=str(uuid.uuid4()),
+            version=0,
+            tags={},
+            module=module,
+            func_qname=function,
+            evaluate_at=ProbeEvaluateTimingForMethod.ENTER,
+            template=message,
+            segments=[LiteralTemplateSegment(message)],
+            take_snapshot=True,
+            limits=DEFAULT_CAPTURE_LIMITS,
+            condition=None,
+            condition_error_rate=0.0,
+            rate=float("inf"),
+        )
+
+
+@attr.s
+class ExitSpanProbe(LogLineProbe):
```

Comment:
do we want to keep exit probe as line probes? or do we want to move all exit span probes to be method probes?